websocketOutportDriver/sender/webSocketSender.go

```diff
@@ -0,0 +1,237 @@
+	}
+	if check.IfNil(args.Uint64ByteSliceConverter) {
+		return nil, ErrNilUint64ByteSliceConverter
+	}
+	if check.IfNil(args.Log) {
+		return nil, ErrNilLogger
+	}
+
+	ws := &webSocketSender{
+		log:                      args.Log,
+		server:                   args.Server,
+		counter:                  0,
+		uint64ByteSliceConverter: args.Uint64ByteSliceConverter,
+		clientsHolder:            NewWebsocketClientsHolder(),
+		acknowledges:             NewAcknowledgesHolder(),
+		withAcknowledge:          args.WithAcknowledge,
+	}
+
+	go ws.start()
+
+	return ws, nil
+}
+
+// AddClient will add the client to internal maps and will also start
+func (w *webSocketSender) AddClient(wss WSConn, remoteAddr string) {
+	client := &webSocketClient{
+		conn:       wss,
+		remoteAddr: remoteAddr,
+	}
+
+	w.clientsHolder.AddClient(client)
+
+	w.acknowledges.AddEntryForClient(remoteAddr)
+
+	if !w.withAcknowledge {
+		return
+	}
+
+	go w.handleReceiveAck(client)
+}
+
+func (w *webSocketSender) handleReceiveAck(client *webSocketClient) {
+	for {
+		mType, message, err := client.conn.ReadMessage()
+		if err != nil {
+			w.log.Error("cannot read message", "remote addr", client.remoteAddr, "error", err)
+			w.clientsHolder.Remove(client.remoteAddr)
+
+			break
+		}
+
+		if mType != websocket.BinaryMessage {
+			w.log.Warn("received message is not binary message", "remote addr", client.remoteAddr, "message type", mType)
+			continue
+		}
+
+		w.log.Debug("received ack", "remote addr", client.remoteAddr, "message", message)
+		counter, err := w.uint64ByteSliceConverter.ToUint64(message)
+		if err != nil {
+			w.log.Warn("cannot decode counter: bytes to uint64",
+				"remote addr", client.remoteAddr,
+				"counter bytes", message,
+				"error", err,
+			)
+			continue
+		}
+
+		w.acknowledges.AddReceivedAcknowledge(client.remoteAddr, counter)
+	}
+}
+
+func (w *webSocketSender) start() {
+	err := w.server.ListenAndServe()
+	if err != nil && !strings.Contains(err.Error(), "http: Server closed") {
+		w.log.Error("could not initialize webserver", "error", err)
+	}
+}
+
+func (w *webSocketSender) sendDataToClients(
+	data []byte,
+	counter uint64,
+) error {
+	numSent := 0
+	var err error
+
+	clients := w.clientsHolder.GetAll()
+	if len(clients) == 0 {
+		return ErrNoClientToSendTo
+	}
+
+	for _, client := range w.clientsHolder.GetAll() {
+		err = w.sendData(data, *client, counter)
+		if err != nil {
+			w.log.Error("couldn't send data to client", "error", err)
+			continue
+		}
+
+		numSent++
+	}
+
+	if numSent == 0 {
+		return fmt.Errorf("data wasn't sent to any client. last known error: %w", err)
+	}
+
+	return nil
+}
+
+func (w *webSocketSender) sendData(
+	data []byte,
+	client webSocketClient,
+	counter uint64,
+) error {
+	if len(data) == 0 {
+		return ErrEmptyDataToSend
+	}
+
+	errSend := client.conn.WriteMessage(websocket.BinaryMessage, data)
+	if errSend != nil {
+		w.log.Warn("could not send data to client", "remote addr", client.remoteAddr, "error", errSend)
```

Comment:
On L96 when you can't read the message from a client you remove it:
```
w.clientsHolder.Remove(client.remoteAddr)
```
Why not do the same here?

Reply:
Otherwise, you come back to this problem: 
https://github.com/ElrondNetwork/elrond-go-core/pull/30#discussion_r748118437

Reply:
done. good suggestion

Reply:
Kind reminder to fix/test this :pray:

Reply:
added a todo for further testing